// @sanity/desk-tool/lib/pane/Editor/Actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactTippy = require("react-tippy");

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _languageSelectComponent = _interopRequireDefault(require("part:@sanity/desk-tool/language-select-component?"));

var _list = _interopRequireDefault(require("part:@sanity/components/validation/list"));

var _chevronDownIcon = _interopRequireDefault(require("part:@sanity/base/chevron-down-icon"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _Editor = _interopRequireDefault(require("../styles/Editor.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Actions extends _react.default.PureComponent {
  renderErrors() {
    var _this$props = this.props,
        onCloseValidationResults = _this$props.onCloseValidationResults,
        onFocus = _this$props.onFocus,
        onToggleValidationResults = _this$props.onToggleValidationResults,
        markers = _this$props.markers,
        showValidationTooltip = _this$props.showValidationTooltip,
        type = _this$props.type;
    var validation = markers.filter(marker => marker.type === 'validation');
    var errors = validation.filter(marker => marker.level === 'error');
    var warnings = validation.filter(marker => marker.level === 'warning');

    if (errors.length === 0 && warnings.length === 0) {
      return null;
    }

    return _react.default.createElement(_reactTippy.Tooltip, {
      arrow: true,
      theme: "light noPadding",
      trigger: "click",
      position: "bottom",
      interactive: true,
      duration: 100,
      open: showValidationTooltip,
      onRequestClose: onCloseValidationResults,
      html: _react.default.createElement(_list.default, {
        truncate: true,
        markers: validation,
        showLink: true,
        isOpen: showValidationTooltip,
        documentType: type,
        onClose: onCloseValidationResults,
        onFocus: onFocus
      })
    }, _react.default.createElement(_default.default, {
      color: "danger",
      bleed: true,
      icon: _warningIcon.default,
      padding: "small",
      onClick: onToggleValidationResults
    }, errors.length, _react.default.createElement("span", {
      style: {
        paddingLeft: '0.5em',
        display: 'flex'
      }
    }, _react.default.createElement(_chevronDownIcon.default, null))));
  }

  render() {
    return _react.default.createElement("div", {
      className: _Editor.default.paneFunctions
    }, _languageSelectComponent.default && _react.default.createElement(_languageSelectComponent.default, null), this.renderErrors());
  }

}

exports.default = Actions;

_defineProperty(Actions, "propTypes", {
  markers: _propTypes.default.arrayOf(_propTypes.default.shape({
    path: _propTypes.default.array
  })).isRequired,
  onCloseValidationResults: _propTypes.default.func.isRequired,
  onFocus: _propTypes.default.func.isRequired,
  onToggleValidationResults: _propTypes.default.func.isRequired,
  showValidationTooltip: _propTypes.default.bool.isRequired,
  type: _propTypes.default.object.isRequired
});
